File: client/src/main/java/com/jamesward/springaimcpdemo/Application.java
package com.jamesward.springaimcpdemo;

import io.modelcontextprotocol.client.McpSyncClient;
import io.modelcontextprotocol.spec.McpSchema;
import org.slf4j.LoggerFactory;
import org.slf4j.event.Level;
import org.springaicommunity.mcp.annotation.McpElicitation;
import org.springaicommunity.mcp.annotation.McpLogging;
import org.springaicommunity.mcp.annotation.McpProgress;
import org.springaicommunity.mcp.annotation.McpSampling;
import org.springframework.boot.ApplicationRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import java.util.List;
import java.util.Map;

@SpringBootApplication
public class Application {

    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }

    @Bean
    public ApplicationRunner applicationRunner(List<McpSyncClient> mcpClients) {
        return args ->
            mcpClients.forEach(client -> {
                System.out.println(client.getServerInfo());
                System.out.println(client.getServerCapabilities());

                // tools
                McpSchema.CallToolResult toolAddResult = client.callTool(new McpSchema.CallToolRequest("add", Map.of("x", 1, "y", 2)));
                System.out.println(toolAddResult);

                // structured content
                McpSchema.CallToolResult toolMultResult = client.callTool(new McpSchema.CallToolRequest("multiply", Map.of("x", 2, "y", 3)));
                Map<?, ?> toolMultContent = (Map<?, ?>) toolMultResult.structuredContent();
                System.out.println("result of multiple = " + toolMultContent.get("result"));

                // sends log message
                McpSchema.CallToolResult toolSubResult = client.callTool(new McpSchema.CallToolRequest("sub", Map.of("x", 4, "y", 1)));
                System.out.println(toolSubResult);

                // sends progress notifications
                McpSchema.CallToolResult toolProgressResult = client.callTool(new McpSchema.CallToolRequest("divide", Map.of("x", 10, "y", 2), Map.of("progressToken", "0")));
                System.out.println(toolProgressResult);

                // elicitation - repeated because only sometimes is something elicited
                for (int i = 0; i < 5; i++) {
                    McpSchema.CallToolResult toolRandomResult = client.callTool(new McpSchema.CallToolRequest("random", Map.of()));
                    System.out.println(toolRandomResult);
                }

                // resources
                System.out.println(client.listResourceTemplates());

                McpSchema.ReadResourceResult resource = client.readResource(new McpSchema.ReadResourceRequest("config://foobar"));
                System.out.println(resource);

                // completions
                McpSchema.CompleteRequest resourceCompletionRequest = new McpSchema.CompleteRequest(
                    new McpSchema.ResourceReference("config://{key}"),
                    new McpSchema.CompleteRequest.CompleteArgument("key", "a")
                );
                McpSchema.CompleteResult resourceCompletion = client.completeCompletion(resourceCompletionRequest);
                System.out.println(resourceCompletion);

                // prompts
                McpSchema.GetPromptResult promptResult = client.getPrompt(new McpSchema.GetPromptRequest("greeting", Map.of("name", "James")));
                System.out.println(promptResult);

                // sampling
                Object loudJoke = client.callTool(new McpSchema.CallToolRequest("loudJoke", Map.of())).content();
                System.out.println(loudJoke);
            });
    }

    @McpLogging(clients = {"demo"})
    public void handleLoggingMessage(McpSchema.LoggingMessageNotification notification) {
        LoggerFactory.getLogger(notification.logger())
            .atLevel(Level.valueOf(notification.level().name()))
            .log(notification.data());
    }

    @McpProgress(clients = {"demo"})
    public void handleProgressNotification(McpSchema.ProgressNotification notification) {
        double percentage = notification.progress() * 100;
        System.out.printf("Progress: %.2f%% - %s%n", percentage, notification.message());
    }

    @McpElicitation(clients = {"demo"})
    public McpSchema.ElicitResult handleElicitationRequest(McpSchema.ElicitRequest request) {
        System.out.println("Server has elicited: " + request.message());
        Map<String, Object> userData = Map.of("number", "12345");
        return new McpSchema.ElicitResult(McpSchema.ElicitResult.Action.ACCEPT, userData);
    }

    @McpSampling(clients = {"demo"})
    public McpSchema.CreateMessageResult handleSampling(McpSchema.CreateMessageRequest request) {
        System.out.println("Server has requested sampling: " + request);
        String response = "You're absolutely right!";
        return McpSchema.CreateMessageResult.builder()
            .role(McpSchema.Role.ASSISTANT)
            .content(new McpSchema.TextContent(response))
            .build();
    }
}

File: client/src/main/kotlin/com/jamesward/springaimcpdemo/Application.kt
package com.jamesward.springaimcpdemo

import io.modelcontextprotocol.client.McpSyncClient
import io.modelcontextprotocol.spec.McpSchema
import org.slf4j.LoggerFactory
import org.slf4j.event.Level
import org.springaicommunity.mcp.annotation.McpElicitation
import org.springaicommunity.mcp.annotation.McpLogging
import org.springaicommunity.mcp.annotation.McpProgress
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean


@SpringBootApplication
class Application {

    @Bean
    fun applicationRunner(mcpClients: List<McpSyncClient>) = ApplicationRunner {
        mcpClients.forEach { client ->
            println(client.serverInfo)
            println(client.serverCapabilities)

            // tools

//            println(client.listTools())
            val toolAddResult = client.callTool(McpSchema.CallToolRequest("add", mapOf("x" to 1, "y" to 2)))
            println(toolAddResult)

            // structured content
            val toolMultResult = client.callTool(McpSchema.CallToolRequest("multiply", mapOf("x" to 2, "y" to 3)))
            val toolMultContent = toolMultResult.structuredContent() as Map<*, *>
            println("result of multiple = ${toolMultContent["result"]}")

            // sends log message
            val toolSubResult = client.callTool(McpSchema.CallToolRequest("sub", mapOf("x" to 4, "y" to 1)))
            println(toolSubResult)

            // sends progress notifications
            val toolProgressResult = client.callTool(McpSchema.CallToolRequest("divide", mapOf("x" to 10, "y" to 2), mapOf("progressToken" to "0")))
            println(toolProgressResult)

            // elicitation - repeated because only sometimes is something elicited
            repeat(5) {
                val toolRandomResult = client.callTool(McpSchema.CallToolRequest("random", emptyMap()))
                println(toolRandomResult)
            }

            // resources

            println(client.listResourceTemplates())

            val resource = client.readResource(McpSchema.ReadResourceRequest("config://foobar"))
            println(resource)

            // completions

            val resourceCompletionRequest = McpSchema.CompleteRequest(McpSchema.ResourceReference("config://{key}"), McpSchema.CompleteRequest.CompleteArgument("key", "a"))
            val resourceCompletion = client.completeCompletion(resourceCompletionRequest)
            println(resourceCompletion)

            // prompts

            val promptResult = client.getPrompt(McpSchema.GetPromptRequest("greeting", mapOf("name" to "James")))
            println(promptResult)
        }
    }

    @McpLogging(clients = ["demo"]) // client id aligns to id in client connection config
    fun handleLoggingMessage(notification: McpSchema.LoggingMessageNotification) {
        LoggerFactory.getLogger(notification.logger())
            .atLevel(Level.valueOf(notification.level.name))
            .log(notification.data())
    }

    @McpProgress(clients = ["demo"])
    fun handleProgressNotification(notification: McpSchema.ProgressNotification) {
        val percentage = notification.progress() * 100
        println(
            String.format(
                "Progress: %.2f%% - %s",
                percentage, notification.message()
            )
        )
    }

    @McpElicitation(clients = ["demo"]) // client id aligns to id in client connection config
    fun handleElicitationRequest(request: McpSchema.ElicitRequest): McpSchema.ElicitResult {
        println("Server has elicited: ${request.message()}")
        val userData = mapOf("number" to "12345")
        return McpSchema.ElicitResult(McpSchema.ElicitResult.Action.ACCEPT, userData)
    }

}

fun main(args: Array<String>) {
    runApplication<Application>(*args)
}
